src/odm_sharing/private/trees.py

```diff
@@ -0,0 +1,408 @@
+            )
+
+        def init_share_node(rule_id: int, org: str, table_nodes: List[Node]
+                            ) -> Node:
+            return Node(
+                rule_id=rule_id,
+                kind=NodeKind.SHARE,
+                value=org,
+                sons=table_nodes
+            )
+
+        orgs = parse_list(ctx, key, min=1)
+        ids = parse_int_list(ctx, val_str, min=1)
+        share_value_nodes = seq(ids).map(get_ctx_node).list()
+        select_nodes = seq(share_value_nodes).filter(is_select).list()
+        table_select_node_ids = get_table_select_nodes(ctx, select_nodes)
+        init_table_node2 = partial(init_table_node, share_value_nodes)
+        table_nodes = seq(table_select_node_ids.items())\
+            .smap(init_table_node2)\
+            .list()
+        share_nodes = seq(orgs)\
+            .map(lambda org: init_share_node(rule_id, org, table_nodes))\
+            .list()
+        root_node_candidate = Node(
+            rule_id=0,
+            kind=NodeKind.ROOT,
+            sons=share_nodes
+        )
+        return root_node_candidate
+    else:
+        assert False, 'not all cases covered'
+
+
+def add_node(ctx: Ctx, rule_id: RuleId, table: str, mode: RuleMode,
+             key: str, op_str: str, val_str: str) -> None:
+    '''initializes and adds a new node to the rule tree
+
+    :param table: table name, may be empty
+    :raises ParseError:
+    '''
+    if table:
+        ctx.node_tables[rule_id].append(table)
+    node = init_node(ctx, rule_id, table, mode, key, op_str, val_str)
+    ctx.nodes[rule_id] = node
+    if len(table) > 0:
+        ctx.table_nodes[table][rule_id] = node
+    if node.kind == NodeKind.ROOT:
+        if not ctx.root:
+            ctx.root = node
+        else:
+            ctx.root = Node(
+                rule_id=0,
+                kind=NodeKind.ROOT,
+                sons=(ctx.root.sons+node.sons)
+            )
+
+
+def validate_schema(ctx: Ctx, rules: List[Rule]) -> None:
+    '''checks that the required rules are present in the schema, or raises
+    ParseError'''
+    if not seq(rules).filter(lambda x: x.mode == RuleMode.SHARE).any():
+        fail(ctx, 'no share-rules in schema')
+    if not seq(rules).filter(lambda x: x.mode == RuleMode.SELECT).any():
+        fail(ctx, 'no select-rules in schema')
+
+
+def validate_rule_tables(ctx: Ctx, rule: Rule, tables: List[str]) -> None:
+    '''checks rule table values
+
+    :param tables: list of table names, or a list with a single empty string
+    (to indicate zero tables)
+    :raises ParseError:
+    '''
+    has = bool(tables[0])
+    should_have = rule.mode in TABLE_MODES
+    if has and not should_have:
+        fail(ctx, f'tables are not allowed for mode {rule.mode}')
+    elif not has and should_have:
+        fail(ctx, f'tables are required for mode {rule.mode}')
+
+
+def parse(rules: Union[Dict[RuleId, Rule], List[Rule]],
```

Comment:
Why can this parameter also take a `List[Rule]`?

Reply:
because rules.load returns a dict which should be accepted by this function, but the function only uses it as a list of rules, so I thought that a list should be accepted as well, in case the user wants to use it that way.

Reply:
Is there a reason `rules.load` returns a dictionary rather than a list of rules? That seems simpler.